enhancements/microshift/y-minus-2-upgrades.md

```diff
@@ -0,0 +1,304 @@
+no skew at all.
+
+If, in the future, MicroShift does need to support multi-node
+deployments there will be many other aspects of deployment and upgrade
+to consider, in addition to the version skew problem. We can envision
+implementing a process similar to what OpenShift uses, where the
+control plane and workers are updated using separate steps. This would
+make the single-node configuration of MicroShift and the multi-node
+configuration mirror the trade-offs of being able to upgrade the
+entire cluster at one time or offering no downtime that are present in
+SNO and HA OCP.
+
+If an upgrade fails, even after a complex data migration, MicroShift's
+rollback process is to discard the new database and restore the old
+version from a backup before continuing. This ensures that an old
+version of the software matches the older database (file format,
+schema, and content).
+
+MicroShift does not automatically create `StorageVersionMigration` CRs
+to trigger data migration. The core kubernetes APIs are safe because
+upstream has committed to not drop any storage versions. CRDs
+installed on top could be more of an issue, but they are installed by
+the end user so it's up to them to track the need for updates.
+
+### Drawbacks
+
+The main drawback to implementing this enhancement is the increased
+test matrix for upgrades. We can automate those tests to minimize the
+impact.
+
+## Test Plan
+
+We will add an automated test to CI to deploy 4.Y-2 and update to 4.Y
+using the latest published packages of 4.Y-2 and testing the "source"
+version (HEAD of the branch or the pull request content) of 4.Y. This
+ensures that every package we build can be continuously upgraded to
+the latest version of the source.
+
+MicroShift's OS support policy is to allow combining each version of
+MicroShift with 1 EUS version of RHEL and the next non-EUS version of
+RHEL. We test upgrades from 4.Y-1 to 4.Y with the same underlying OS
+and also moving from the EUS version to non-EUS version. The aspects
+of testing the OS support during upgrades are orthogonal to the work
+for this enhancement, however, and should not require additional
+expansion of the test matrix, either in CI or by QE.
+
+The QE team will need to perform similar tests using the 4.Y-2 and 4.Y
+packages built by the release team. Their test plan includes:
+
+1. The scope of QE testing will be confined to upgrades of RHEL and
+   rpm-ostree hosts deployed with RPMs built by ART (nightly, EC, RC
+   and after release z-streams).
+1. Both x86_64 and aarch64 architecture will be covered.
+   1. Not every combination listed below will be tested on both
+      architectures.
+   1. A mix of Intel and ARM hosts will be used
+1. Rpm-ostree based hosts
+   1. Initial cluster bring-up will be a mix of deployments from ISO
+      installer and rpm-ostree upgrades from a bare RHEL host
+   1. The following upgrade paths will be covered:
+      1. RHEL 9.2 / 4.14.latest → RHEL 9.4 / 4.16.Z
```

Comment:
Is "4.14.latest" going to start with the 4.14. z-stream version when the 4.16.0 version is released? So can users update from a z-stream prior, down to 4.14.0? Or does the Y-version check in reality ignore the z-stream skew?

Reply:
In CI we test by starting from the latest from a z-stream. That ensures there is at least _some_ path to update, and since it is extremely unlikely that we would introduce a change into the z-stream that would break updates the benefit of testing every combination of versions from the z-stream to the new release does not make the effort involved worth it.

@jogeo, I assume QE is taking a similar stance?